migrator/migrator.go

```diff
@@ -523,13 +522,105 @@ func (m Migrator) MigrateColumn(value interface{}, field *schema.Field, columnTy
 		}
 	}
 
-	if alterColumn && !field.IgnoreMigration {
-		return m.DB.Migrator().AlterColumn(value, field.DBName)
+	if alterColumn {
+		if err := m.DB.Migrator().AlterColumn(value, field.DBName); err != nil {
+			return err
+		}
+	}
+
+	if err := m.MigrateColumnUnique(value, field, columnType); err != nil {
+		return err
 	}
 
 	return nil
 }
 
+func (m Migrator) MigrateColumnUnique(value interface{}, field *schema.Field, columnType gorm.ColumnType) error {
+	unique, ok := columnType.Unique()
+	if !ok || field.PrimaryKey {
+		return nil // skip primary key
+	}
+
+	queryTx := m.DB.Session(&gorm.Session{})
+	execTx := queryTx
+	if m.DB.DryRun {
+		queryTx.DryRun = false
+		execTx = m.DB.Session(&gorm.Session{Logger: &printSQLLogger{Interface: m.DB.Logger}})
+	}
+
+	return m.RunWithValue(value, func(stmt *gorm.Statement) error {
+		// We're currently only receiving boolean values on `Unique` tag,
+		// so the UniqueConstraint name is fixed
+		constraint := m.DB.NamingStrategy.UniqueName(stmt.Table, field.DBName)
+		if m.UniqueAffectedByUniqueIndex {
+			if unique {
+				// Clean up redundant unique indexes
+				indexes, _ := queryTx.Migrator().GetIndexes(value)
+				for _, index := range indexes {
+					if uni, ok := index.Unique(); !ok || !uni {
+						continue
+					}
+					if columns := index.Columns(); len(columns) != 1 || columns[0] != field.DBName {
+						continue
+					}
+					if name := index.Name(); name == constraint || name == field.UniqueIndex {
+						continue
+					}
+					if err := execTx.Migrator().DropIndex(value, index.Name()); err != nil {
+						return err
+					}
+				}
```

Comment:
Because Unique is affected by UniuqeIndex,
I had to clear redundant indexes here.

But this makes different `Dialector` behave differently.
For example, in unit tests, mysql behaves differently from other dialectors.
See https://github.com/go-gorm/gorm/pull/6386/files#diff-23864390402317a89bc6a5197a8144403c082c4e2a3d0132bac131454831b5f9R1717-R1760

Reply:
Just watching this unfold as a spectator eagerly awaiting the outcome. We are currently locked on v1.24.2 because upgrading that version breaks our migrations miserably. Investigation has shown the break is due an issue that this PR appears to fix. Thank you both!